test/tangent_types/notimplemented.jl

```diff
@@ -6,76 +6,51 @@
-        @test muladd(ni, ni2, z) === ni
-        @test muladd(ni, ni2, ZeroTangent()) === ni
-        @test muladd(ni, y, ni2) === ni
-        @test muladd(ni, ZeroTangent(), ni2) === ni2
-        @test muladd(x, ni, z) === ni
-        @test muladd(ZeroTangent(), ni, z) == z
-        @test muladd(x, ni, ZeroTangent()) === ni
-        @test muladd(ZeroTangent(), ni, ZeroTangent()) == ZeroTangent()
-        @test muladd(x, ni, ni2) === ni
-        @test muladd(ZeroTangent(), ni, ni2) === ni2
-        @test muladd(x, y, ni) === ni
-        @test muladd(ZeroTangent(), y, ni) === ni
-        @test muladd(x, ZeroTangent(), ni) === ni
-        @test muladd(ZeroTangent(), ZeroTangent(), ni) === ni
-        @test ni + rand() === ni
-        @test ni + ZeroTangent() === ni
-        @test ni + NoTangent() === ni
-        @test ni + true === ni
-        @test ni + @thunk(x^2) === ni
-        @test rand() + ni === ni
-        @test ZeroTangent() + ni === ni
-        @test NoTangent() + ni === ni
-        @test true + ni === ni
-        @test @thunk(x^2) + ni === ni
+
+        # addition
+        for a in (rand(), NoTangent(), ZeroTangent(), true, @thunk(x^2))
+            @test ni + a === ni
+            @test a + ni === ni
+        end
         @test ni + ni2 === ni
-        @test ni * rand() === ni
-        @test ni * ZeroTangent() == ZeroTangent()
-        @test ZeroTangent() * ni == ZeroTangent()
-        @test dot(ni, ZeroTangent()) == ZeroTangent()
-        @test dot(ZeroTangent(), ni) == ZeroTangent()
+        @test ni2 + ni === ni2
+
+        # multiplication and dot product
+        for a in (rand(), true, @thunk(x^2))
+            @test ni * a === ni
+            @test a * ni === ni
+            @test dot(ni, a) === ni
+            @test dot(a, ni) === ni
+        end
+        for a in (NoTangent(), ZeroTangent())
+            @test ni * a === a
+            @test a * ni === a
+            @test dot(ni, a) === a
+            @test dot(a, ni) === a
+        end
+        @test ni * ni2 === ni
+        @test ni2 * ni === ni2
+        @test dot(ni, ni2) === ni
+        @test dot(ni2, ni) === ni2
+
+        # broadcasting
         @test ni .* rand() === ni
+        @test rand() .* ni === ni
         @test broadcastable(ni) isa Ref{typeof(ni)}
 
         # unsupported operations
         E = ChainRulesCore.NotImplementedException
         @test_throws E +ni
         @test_throws E -ni
-        @test_throws E ni - rand()
-        @test_throws E ni - ZeroTangent()
-        @test_throws E ni - NoTangent()
-        @test_throws E ni - true
-        @test_throws E ni - @thunk(x^2)
-        @test_throws E rand() - ni
-        @test_throws E ZeroTangent() - ni
-        @test_throws E NoTangent() - ni
-        @test_throws E true - ni
-        @test_throws E @thunk(x^2) - ni
+        for a in (rand(), NoTangent(), ZeroTangent(), true, @thunk(x^2))
+            @test_throws E ni - a
+            @test_throws E a - ni
```

Comment:
Seems funky that adding is fine but subtracting throws.

Like I guess it's fine -- adding tangents is what happens in AD, subtracting vis not.

Subtracting happen in gradient descent.

Ok I am convinced this is the behaviour we want.
Do we have a comment saying as such?

Reply:
I'll add a comment.